### PR TITLE
#291-Make-404-page-content-managed: Wired rendering host that queries headless CMS and finds if item exists before displaying (404) item

### PR DESCRIPTION
**Pull request by navancommits**
Rendering host that is able to detect when to display cms-based 404 page

## Description / Motivation

Displays content-managed item instead of static page
Issue#291 is addressed

## How Has This Been Tested?

Tested locally

https://navansitecorenotes.blogspot.com/2023/12/create-aspnet-core-custom-rendering.html

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read the Contributing guide.
- [x] My code/comments/docs fully adhere to the Code of Conduct.
- [x] My change is a code change.
- [ ] My change is a documentation change and there are NO other updates required.

### CODE REVIEW

File: src/Project/MvpSite/rendering/Middleware/CustomGraphQlLayoutServiceHandler.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+        public CustomGraphQlLayoutServiceHandler(IConfiguration configuration, IGraphQLRequestBuilder graphQLRequestBuilder, IGraphQLClientFactory graphQLClientFactory)
+        {
+            _graphQLRequestBuilder = graphQLRequestBuilder;
+            _graphQLClientFactory = graphQLClientFactory;
+            _configuration = configuration.GetSection(MvpSiteSettings.Key).Get<MvpSiteSettings>();
+        }
+
+        public async Task<bool> CheckLayoutExists(
+          SitecoreLayoutRequest layoutrequest)
```
Comment (robearlam) [inline src/Project/MvpSite/rendering/Middleware/CustomGraphQlLayoutServiceHandler.cs:25]: Remove new line here, also variable is incorrectly cased, it should he `layoutRequest`

File: src/Project/MvpSite/rendering/Mvp.Project.MvpSite.Rendering.csproj [hunk 2]
```diff
@@ -23,6 +23,7 @@
 
 	<ItemGroup>
 	  <Content Remove="compilerconfig.json" />
+	  <Content Remove="Views\Shared\NotFound.cshtml" />
```
Comment (robearlam) [inline src/Project/MvpSite/rendering/Mvp.Project.MvpSite.Rendering.csproj:26]: This shouldn't be needed as we're hosting in PaaS not IaaS, the app is deployed fresh each time so there will be no legacy files to remove

File: src/Project/MvpSite/rendering/Middleware/CustomRenderingMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+            httpContext.Items.Add((object)nameof(RenderingEngineMiddleware), (object)null);
+
+            await _next(httpContext).ConfigureAwait(continueOnCapturedContext: false);
+        }
+
+        private async Task<SitecoreLayoutResponse> GetSitecoreLayoutResponse(HttpContext httpContext)
+        {
+            //intercept for not found logic - start
+            CustomGraphQlLayoutServiceHandler customGraphQlLayoutServiceHandler = new(_configuration,_graphQLRequestBuilder, _graphQLClientFactory);
```
Comment (robearlam) [inline src/Project/MvpSite/rendering/Middleware/CustomRenderingMiddleware.cs:77]: Don't use `new` to create instances like this, it should be injected into the class ctor via DI

File: src/Project/MvpSite/rendering/Middleware/CustomRenderingMiddleware.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            SitecoreLayoutRequest sitecoreLayoutRequest = _requestMapper.Map(httpContext.Request);
+            bool retVal = await customGraphQlLayoutServiceHandler.CheckLayoutExists(sitecoreLayoutRequest);
+            if (!retVal)  httpContext.Request.Path = _configuration.GetSection(MvpSiteSettings.Key).Get<MvpSiteSettings>().NotFoundPage;
+            //intercept for not found logic - end
+
+            sitecoreLayoutRequest = _requestMapper.Map(httpContext.Request);
+            return await _layoutService.Request(sitecoreLayoutRequest).ConfigureAwait(continueOnCapturedContext: false);
```
Comment (robearlam) [inline src/Project/MvpSite/rendering/Middleware/CustomRenderingMiddleware.cs:84]: The logic here isn't right, adding `awaitustomGraphQlLayoutServiceHandler.CheckLayoutExists(sitecoreLayoutRequest);` is causing an extra GQL request to be made to edge for every page request.

This needs to be changed to make the initial page request as normal, and then we should interrogate the response and only then make and extra request for the 404 page when required. That would remove the extra overhead being introduced here.

**Review by robearlam — CHANGES_REQUESTED** [review]
Hey @navancommits, this is looking good, There are a small formatting changes that need to happen, 

The main issue is the introduction of the `CheckLayoutExists` call though, which will add an extra call to edge for every page request. As listed is the comment below, this needs to be addressed before this can be merged in.

**Comment by navancommits** [comment]
Fixed...

**Review by robearlam — APPROVED** [review]

**Pull request merged**